fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/JobsServlet.java

```diff
@@ -431,10 +440,85 @@ protected void handleResultsResponse(
             );
             return Observable.error(responseException);
         }
+
+        if (jobsApiRequest.getPaginationParameters().isPresent()) {
+            Pagination<Result> pages = new AllPagesPagination<>(
+                    preResponse.getResultSet(),
+                    jobsApiRequest.getPaginationParameters().get()
+            );
+            responseContext = paginateResponseContext(responseContext, jobsApiRequest, pages);
+            PreResponse paginatedPreResponse = new PreResponse(
+                    new ResultSet(pages.getPageOfData(), preResponse.getResultSet().getSchema()),
+                    responseContext
+            );
+            return Observable.just(paginatedPreResponse);
+        }
```

Comment:
You can make this a bit cleaner (well, I think it's cleaner, YMMV) if you take advantage of the `map` function on `Optional`:

```java
return jobsApiRequest.getPaginationParameters()
        .map(paginationParameters -> new AllPagesPagination<>(preResponse.getResultSet(), paginationParameters))
        .map(pages -> new PreResponse(
                new ResultSet(pages.getPageOfData(), preResponse.getResultSet().getSchema()),
                paginateResponseContext(responseContext, jobsApiRequest, pages)
        ))
        .map(Observable::just)
        .orElse(Observable.just(preResponse));
```
Given an `Optional<T>`, `map` takes a `Function<T, U>`, and returns an `Optional<U>`. If the original `Optional` is empty, then the `map` returns the empty `Optional`. If the original `Optional` is not empty, however, then it applies the `Function<T, U>` to the value contained in the `Optional` and returns a new `Optional` wrapping the result.

Meanwhile, `orElse` takes a default value. If the `Optional` it is applied to is empty, then it returns the default value. Otherwise, it returns the value wrapped in the `Optional`.

This allows us to avoid a fat if statement. I don't like if-statements because as you're looking through the if-block, you have to keep in your head what the check was that triggered entering the if-statement. So instead of thinking "OK, suppose the value is in the `Optional`. Then we do X and return Y. Ok, we're leaving the if-block, so if we the `Optional` is empty then we return Z." This instead allows us to think "OK do X to transform Y' into Y, and return it, defaulting to Z."